src/garage/torch/algos/ddpg.py

```diff
@@ -0,0 +1,300 @@
+        self.qf_optimizer = optimizer(
+            self.qf._nn_module.parameters(), lr=self.qf_lr)
+
+    def train_once(self, itr, paths):
+        """Perform one iteration of training."""
+        paths = self.process_samples(itr, paths)
+
+        epoch = itr / self.n_epoch_cycles
+
+        self.episode_rewards.extend([
+            path for path, complete in zip(paths['undiscounted_returns'],
+                                           paths['complete']) if complete
+        ])
+        self.success_history.extend([
+            path for path, complete in zip(paths['success_history'],
+                                           paths['complete']) if complete
+        ])
+
+        last_average_return = np.mean(self.episode_rewards)
+        for train_itr in range(self.n_train_steps):
+            if self.replay_buffer.n_transitions_stored >= self.min_buffer_size:  # noqa: E501
+                self.evaluate = True
+                qf_loss, y, q, policy_loss = self.torch_to_np(
+                    self.optimize_policy(itr, paths))  # pylint: disable=all
+
+                self.episode_policy_losses.append(policy_loss)
+                self.episode_qf_losses.append(qf_loss)
+                self.epoch_ys.append(y)
+                self.epoch_qs.append(q)
+
+        if itr % self.n_epoch_cycles == 0:
+            logger.log('Training finished')
+
+            if self.evaluate:
+                tabular.record('Epoch', epoch)
+                tabular.record('AverageReturn', np.mean(self.episode_rewards))
+                tabular.record('StdReturn', np.std(self.episode_rewards))
+                tabular.record('Policy/AveragePolicyLoss',
+                               np.mean(self.episode_policy_losses))
+                tabular.record('QFunction/AverageQFunctionLoss',
+                               np.mean(self.episode_qf_losses))
+                tabular.record('QFunction/AverageQ', np.mean(self.epoch_qs))
+                tabular.record('QFunction/MaxQ', np.max(self.epoch_qs))
+                tabular.record('QFunction/AverageAbsQ',
+                               np.mean(np.abs(self.epoch_qs)))
+                tabular.record('QFunction/AverageY', np.mean(self.epoch_ys))
+                tabular.record('QFunction/MaxY', np.max(self.epoch_ys))
+                tabular.record('QFunction/AverageAbsY',
+                               np.mean(np.abs(self.epoch_ys)))
+                if self.input_include_goal:
+                    tabular.record('AverageSuccessRate',
+                                   np.mean(self.success_history))
+
+            if not self.smooth_return:
+                self.episode_rewards = []
+                self.episode_policy_losses = []
+                self.episode_qf_losses = []
+                self.epoch_ys = []
+                self.epoch_qs = []
+
+            self.success_history.clear()
+
+        return last_average_return
+
+    def optimize_policy(self, itr, samples_data):
```

Comment:
I think `samples_data` is not used at all. Should we move the `replay_buffer.sample()` out of this function and use that as samples_data?